File: edexOsgi/com.raytheon.uf.edex.plugin.mesowest/src/com/raytheon/uf/edex/plugin/mesowest/MESOWestDecoder.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.uf.edex.plugin.mesowest;

import java.util.HashMap;
import java.util.Map;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import com.raytheon.edex.exception.DecoderException;
import com.raytheon.uf.common.dataplugin.PluginDataObject;
import com.raytheon.uf.common.dataplugin.PluginException;
import com.raytheon.uf.edex.decodertools.core.IDecoderInput;
import com.raytheon.uf.edex.plugin.mesowest.common.MESOWestRecord;
import com.raytheon.uf.edex.plugin.mesowest.decoder.MESOWestConstants;
import com.raytheon.uf.edex.plugin.mesowest.decoder.MESOWestParser;

/**
 * TODO Add Description
 * 
 * <pre>
 *
 * SOFTWARE HISTORY
 *
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Mar 3, 2009            jkorman     Initial creation
 *
 * </pre>
 *
 * @author jkorman
 * @version 1.0	
 */

public class MESOWestDecoder {

    private Log logger = LogFactory.getLog(getClass());

    private String pluginName = "mesowest";

    private Map<String,MESOWestParser> parserMap = new HashMap<String,MESOWestParser>();
    
    /**
     * 
     * @param name
     */
    public MESOWestDecoder(String name) {
        this.pluginName = name;
    }

    /**
     * Get the next decoded data record.
     * 
     * @return One record of decoded data.
     * @throws DecoderException
     *             Thrown if no data is available.
     */
    public PluginDataObject[] decode(IDecoderInput input) {

        PluginDataObject[] decodedData = null;
        String traceId = null;
        
        logger.debug("MESOWestDecoder.decode()");
        
        if(input != null) {
            traceId = (String) input.getProperty(MESOWestConstants.TRACEID);

            MESOWestRecord record = null;
            
            String type = input.getProperty(MESOWestConstants.K_DATATYPE);
            if(MESOWestConstants.T_PARMHEADER.equals(type)) {
                parserMap.put(input.getProperty("uuid"),new MESOWestParser(input.getReport()));
                logger.debug("Created parser ");
            } else if (MESOWestConstants.T_LASTITEM.equals(type)) {
                parserMap.remove(input.getProperty("uuid"));
                logger.debug("Destroyed parser ");
            } else {
                MESOWestParser parser = parserMap.get(input.getProperty("uuid"));
                if(parser != null) {
                    if(input.getReport().length() > 10) {
                        if((record = parser.decode(input.getReport())) != null) {
                            record.setPluginName(pluginName);
                            record.setTraceId(traceId);
                            record.setObsText(input.getReport() + "\n");
                        }
                    }
                } else {
                    logger.error("Unexpected data in data stream");
                }
            }
            
            try {
                if (record != null) {
                    logger.info("Decoded obs " + record.getStationId());
                    
                    try {
                        record.constructDataURI();
                    } catch (PluginException e) {
                        throw new DecoderException("Unable to construct dataURI", e);
                    }

                    decodedData = new PluginDataObject[] { record };
                } else {
                    logger.info(String.format("%s - Decoded no obs", traceId));
                }
            } catch (Exception e) {
                logger.error("Error in MESOWestDecoder", e);
            } finally {
                if(decodedData == null) {
                    decodedData = new PluginDataObject[0];
                }
            }
        } else {
            logger.error("null input data in MESOWestDecoder");
        }

        return decodedData;
    }
    
    /**
     * @return the pluginName
     */
    public String getPluginName() {
        return pluginName;
    }

}

File: edexOsgi/com.raytheon.uf.edex.plugin.mesowest/MesowestDecoder.py
##
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
# 
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
# 
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
# 
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
##
import sys, os, time, re, string, getopt
import copy
import LogStream
import pupynere
from com.raytheon.uf.common.pointdata.spatial import SurfaceObsLocation
from java.lang import Double
from java.lang import Integer

class MesowestDecoder():

    def __init__(self, text=None, filePath=None):
        
        self._parameters = ['altimeter', 'dewpoint', 'relHumidity', 'stationId', 'rawMessage', 'temperature', 'windDir', 'windGust', 'windSpeed', 'precipAccum']
        self._mappings = ['altimeter', 'dwpt', 'humidity', 'stationId', 'obsText', 'temp', 'windDirection', 'windGust', 'windSpeed', 'precip']

        
        #to ensure time calls are based on Zulu
        os.environ['TZ'] = "GMT0"
        
        self._deleteAfterProcessing = 0
        if filePath is None:
            self._incomingFilename = None
        else:
            self._incomingFilename = filePath

    def decode(self):
        fillValue = None 
        missing = None
        mesowestReports = list()
        obsList = dict()
        net = pupynere.netcdf_file(self._incomingFilename,"r")
        records = net._recs
        for x in range(records):
            stationId = ''.join(net.variables['stationId'][x])
            reportTime = net.variables['reportTime'][x] * 1000
            if obsList.has_key(reportTime):
                if stationId in obsList[reportTime]:
                    continue
                else:
                    obsList[reportTime].append(stationId)
            else:
                obsList[reportTime] = list()
                obsList[reportTime].append(stationId)
            mesowestReport = dict()
            for paramNum in range(len(self._parameters)):
                param = self._parameters[paramNum]
                mapping = self._mappings[paramNum]
                variable = net.variables[param]
                # Set the missing value if available
                try:
                    missing = variable.missing_value
                except AttributeError:
                    missing = None
                # Set the fill value if available
                try:
                    fillValue = variable._FillValue
                except AttributeError:
                    fillValue = None
                if len(variable.shape) == 1:
                    if fillValue != None and variable[x] == fillValue:
                        mesowestReport[mapping] = Double(-9999.0)
                        continue
                    if missing != None:
                        #  Odd case where certain values are -764 in the netcdf
                        if variable[x] == missing or variable[x] < -600:
                            mesowestReport[mapping] = Double(-9999.0)
                            continue
                    mesowestReport[mapping] = Double.valueOf(str(variable[x]))
                elif len(variable.shape) > 1 and variable.shape[1] > 3:
                    mesowestReport[mapping] = ''.join(variable[x])
                    
            location = SurfaceObsLocation()
            lat = Double.valueOf(str(net.variables['latitude'][x]))
            lon = Double.valueOf(str(net.variables['longitude'][x]))
            elv = Double.valueOf(str(net.variables['elevation'][x]))
            location.assignLocation(lat.doubleValue(),lon.doubleValue());
            location.setElevation(Integer(elv.intValue()));
            location.setStationId(stationId)
            mesowestReport['location'] = location
            mesowestReport['timeObs'] = long(reportTime)
            mesowestReport['dataTime'] = mesowestReport['timeObs']
            mesowestReport['networkType'] = ''.join(net.variables['stationType'][x])
            mesowestReports.append(mesowestReport)
        net.close()
        return mesowestReports


    def _usage(self):
        #Prints out usage information if started without sufficient command
        #line arguments.
        s =  "This tool is not meant to be run from the command line."
        print s
        LogStream.logProblem(s)



def main():
    try:
        LogStream.logEvent("MesowestDecoder Starting")        
        decoder = MesowestDecoder()
        decoder.decode()
        decoder = None
        LogStream.logEvent("MesowestDecoder Finished")
    except:
        LogStream.logProblem("Caught Exception: ", LogStream.exc())
        sys.exit(1)

if __name__ == "__main__":
    main()
    sys.exit(0)
